xde/src/xde.rs

```diff
@@ -2245,6 +2272,58 @@ fn dump_tcp_flows_hdlr(
     api::dump_tcp_flows(&dev.port, &req)
 }
 
+#[no_mangle]
+fn set_dhcp_params_hdlr(env: &mut IoctlEnvelope) -> Result<NoResp, OpteError> {
+    let req: SetDhcpParamsReq = env.copy_in_req()?;
+    let devs = unsafe { xde_devs.read() };
+    let mut iter = devs.iter();
+    let dev = match iter.find(|x| x.devname == req.port_name) {
+        Some(dev) => dev,
+        None => return Err(OpteError::PortNotFound(req.port_name)),
+    };
+
+    let state = get_xde_state();
+    let dhcp_cfg = {
+        let map_locked = state.dhcp_map.read();
+        if let Some(dhcp) = map_locked.get(&dev.vni) {
+            dhcp.clone()
+        } else {
+            return Err(OpteError::PortNotFound(req.port_name));
+        }
+    };
+    let mut dhcp_locked = dhcp_cfg.write();
+    *dhcp_locked = req.data;
```

Comment:
I'm not sure `{Dump,Set}DhcpParams` work as expected if `dhcp_map` is keyed on VNI's:

1. Add port 1 with VNI 100. `dhcp_map[100]` will be the Google/Cloudflare `DhcpCfg` and that's also what port 1 gets configured with.
2. SetDhcpParams on port 1 with a new `DhcpCfg`. This grabs `dhcp_map[100]` and updates the `Arc<KwRwLock<DhcpCfg>>` also held by port 1. All good so far.

But consider if a new port gets added in between on the same VPC and thus same VNI. `dhcp_map[100]` gets overridden with a new base `DhcpCfg` that's passed to port 2. Now when the SetDhcpParams request happens, even though we wanted port 1, we're actually updating the state associated with port 2.

Reply:
```console
# create port 1
$ pfexec ./target/debug/opteadm create-xde vm_foo0 [...]

$ pfexec ./target/debug/opteadm list-ports
LINK                             MAC ADDRESS              IPv4 ADDRESS     EXTERNAL IPv4    IPv6 ADDRESS                             EXTERNAL IPv6                            STATE
vm_foo0                          A8:40:25:FF:00:00        10.0.0.10        192.168.1.20     None                                     None                                     running

$ pfexec ./target/debug/opteadm dump-dhcp-params -p vm_foo0
DHCP Configuration
======================================================================
Hostname: <unset>
Domain: <unset>
Search Domains: <unset>
IPv4 DNS Servers:
- 8.8.8.8
- 1.1.1.1
IPv6 DNS Servers:
- 2606:4700:4700::1111
- 2606:4700:4700::1001
- 2001:4860:4860::8888
- 2001:4860:4860::8844

$ pfexec ./target/debug/opteadm create-xde vm_bar0 [...]

$ pfexec ./target/debug/opteadm list-ports
LINK                             MAC ADDRESS              IPv4 ADDRESS     EXTERNAL IPv4    IPv6 ADDRESS                             EXTERNAL IPv6                            STATE
vm_foo0                          A8:40:25:FF:00:00        10.0.0.10        192.168.1.20     None                                     None                                     running
vm_bar0                          A8:40:25:FF:00:01        10.0.0.11        192.168.1.21     None                                     None                                     running

$ pfexec ./target/debug/opteadm dump-dhcp-params -p vm_bar0
DHCP Configuration
======================================================================
Hostname: <unset>
Domain: <unset>
Search Domains: <unset>
IPv4 DNS Servers:
- 8.8.8.8
- 1.1.1.1
IPv6 DNS Servers:
- 2606:4700:4700::1111
- 2606:4700:4700::1001
- 2001:4860:4860::8888
- 2001:4860:4860::8844

# try updating dhcp config for `vm_foo0` port
$ pfexec ./target/debug/opteadm set-dhcp-params -p vm_foo0 --hostname foo

$ pfexec ./target/debug/opteadm dump-dhcp-params -p vm_foo0
DHCP Configuration
======================================================================
Hostname: foo
Domain: <unset>
Search Domains: <unset>
IPv4 DNS Servers: <unset>
IPv6 DNS Servers: <unset>

$ pfexec ./target/debug/opteadm dump-dhcp-params -p vm_bar0
DHCP Configuration
======================================================================
Hostname: foo
Domain: <unset>
Search Domains: <unset>
IPv4 DNS Servers: <unset>
IPv6 DNS Servers: <unset>
```